File: resources/shell/CursorReset.java
package com.moonciki.cursorsekiro;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.TimeUnit;
import org.json.JSONObject;
import java.util.UUID;

/**
 * Cursor设备ID修改工具 - Java版本
 * 用于重置Cursor编辑器的设备标识
 */
public class CursorReset {
    // ANSI颜色代码
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String BLUE = "\u001B[34m";
    private static final String RESET = "\u001B[0m";

    // 文件路径
    private static final String APPDATA = System.getenv("APPDATA");
    private static final String STORAGE_FILE = APPDATA + "\\Cursor\\User\\globalStorage\\storage.json";
    private static final String BACKUP_DIR = APPDATA + "\\Cursor\\User\\globalStorage\\backups";
    private static final String CURSOR_PATH = System.getenv("LOCALAPPDATA") + "\\Programs\\cursor";

    /**
     * 主方法
     */
    public static void main(String[] args) {
        try {
            // 设置控制台编码为UTF-8
            System.setProperty("file.encoding", "UTF-8");
            
            // 检查管理员权限
            if (!isAdmin()) {
                System.out.println(RED + "[错误]" + RESET + " 请以管理员身份运行此程序");
                restartAsAdmin();
                System.exit(1);
            }

            // 检查Cursor版本
            checkCursorVersion();

            // 关闭Cursor进程
            killCursorProcesses();

            // 创建备份目录
            createBackupDirectory();

            // 备份当前配置
            backupConfiguration();

            // 生成新的ID
            String machineId = generateMachineId();
            String macMachineId = UUID.randomUUID().toString();
            String devDeviceId = UUID.randomUUID().toString();
            String sqmId = "{" + UUID.randomUUID().toString().toUpperCase() + "}";

            // 更新配置文件
            updateConfiguration(machineId, macMachineId, devDeviceId, sqmId);

            // 更新注册表
            updateMachineGuid();

            // 显示结果
            displayResults(machineId, macMachineId, devDeviceId, sqmId);

            // 询问是否禁用自动更新
            handleAutoUpdate();

        } catch (Exception e) {
            System.out.println(RED + "[错误]" + RESET + " 程序执行失败: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * 以管理员权限重启程序
     */
    private static void restartAsAdmin() {
        try {
            String javaBin = System.getProperty("java.home") + File.separator + "bin" + File.separator + "java";
            String currentJar = new File(CursorReset.class.getProtectionDomain().getCodeSource().getLocation().toURI()).getPath();
            
            ProcessBuilder pb = new ProcessBuilder(
                "cmd", "/c",
                "powershell", "Start-Process",
                "-FilePath", javaBin,
                "-ArgumentList", "'-jar','" + currentJar + "'",
                "-Verb", "RunAs"
            );
            pb.start();
        } catch (Exception e) {
            System.out.println(RED + "[错误]" + RESET + " 无法以管理员权限重启: " + e.getMessage());
        }
    }

    /**
     * 检查是否具有管理员权限
     */
    private static boolean isAdmin() {
        try {
            ProcessBuilder pb = new ProcessBuilder("cmd.exe", "/c", "net session");
            pb.redirectErrorStream(true);
            Process p = pb.start();
            p.waitFor();
            return p.exitValue() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * 检查Cursor版本
     */
    private static void checkCursorVersion() {
        try {
            Path packagePath = Paths.get(CURSOR_PATH, "resources", "app", "package.json");
            if (Files.exists(packagePath)) {
                String content = new String(Files.readAllBytes(packagePath));
                JSONObject json = new JSONObject(content);
                System.out.println(GREEN + "[信息]" + RESET + " 当前安装的 Cursor 版本: v" + json.getString("version"));
            } else {
                System.out.println(YELLOW + "[警告]" + RESET + " 无法检测到 Cursor 版本");
            }
        } catch (Exception e) {
            System.out.println(RED + "[错误]" + RESET + " 获取 Cursor 版本失败: " + e.getMessage());
        }
    }

    /**
     * 关闭Cursor进程
     */
    private static void killCursorProcesses() throws Exception {
        System.out.println(GREEN + "[信息]" + RESET + " 检查 Cursor 进程...");
        ProcessBuilder pb = new ProcessBuilder("taskkill", "/F", "/IM", "Cursor.exe");
        pb.redirectErrorStream(true);
        Process p = pb.start();
        p.waitFor(5, TimeUnit.SECONDS);
    }

    /**
     * 创建备份目录
     */
    private static void createBackupDirectory() throws IOException {
        Files.createDirectories(Paths.get(BACKUP_DIR));
    }

    /**
     * 备份配置文件
     */
    private static void backupConfiguration() throws IOException {
        Path sourcePath = Paths.get(STORAGE_FILE);
        if (Files.exists(sourcePath)) {
            String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
            Path backupPath = Paths.get(BACKUP_DIR, "storage.json.backup_" + timestamp);
            Files.copy(sourcePath, backupPath, StandardCopyOption.REPLACE_EXISTING);
        }
    }

    /**
     * 生成新的MachineId
     */
    private static String generateMachineId() {
        String prefix = "auth0|user_";
        byte[] randomBytes = new byte[32];
        new Random().nextBytes(randomBytes);
        StringBuilder sb = new StringBuilder();
        for (byte b : prefix.getBytes(StandardCharsets.UTF8)) {
            sb.append(String.format("%02x", b));
        }
        for (byte b : randomBytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    /**
     * 更新配置文件，使用更安全的方式处理JSON
     */
    private static void updateConfiguration(String machineId, String macMachineId,
                                         String devDeviceId, String sqmId) throws IOException {
        Path configPath = Paths.get(STORAGE_FILE);
        Path tempFile = null;
        
        try {
            // 创建临时文件
            tempFile = Files.createTempFile("cursor_config_", ".tmp");
            
            // 读取原始配置
            String content = new String(Files.readAllBytes(configPath), StandardCharsets.UTF8);
            JSONObject config = new JSONObject(content);
            
            // 备份原始值
            JSONObject backup = new JSONObject();
            backup.put("telemetry.machineId", config.optString("telemetry.machineId"));
            backup.put("telemetry.macMachineId", config.optString("telemetry.macMachineId"));
            backup.put("telemetry.devDeviceId", config.optString("telemetry.devDeviceId"));
            backup.put("telemetry.sqmId", config.optString("telemetry.sqmId"));
            
            // 更新配置
            config.put("telemetry.machineId", machineId);
            config.put("telemetry.macMachineId", macMachineId);
            config.put("telemetry.devDeviceId", devDeviceId);
            config.put("telemetry.sqmId", sqmId);
            
            // 先写入临时文件
            Files.write(tempFile, config.toString(2).getBytes(StandardCharsets.UTF8));
            
            // 验证JSON格式
            new JSONObject(new String(Files.readAllBytes(tempFile), StandardCharsets.UTF8));
            
            // 替换原文件
            Files.move(tempFile, configPath, StandardCopyOption.REPLACE_EXISTING);
            
        } catch (Exception e) {
            System.out.println(RED + "[错误]" + RESET + " 更新配置失败: " + e.getMessage());
            // 如果有临时文件，清理它
            if (tempFile != null) {
                try {
                    Files.deleteIfExists(tempFile);
                } catch (IOException ignored) {}
            }
            throw e;
        }
    }

    /**
     * 改进的MachineGuid更新方法
     */
    private static void updateMachineGuid() throws Exception {
        String guid = UUID.randomUUID().toString();
        String registryPath = "SOFTWARE\\Microsoft\\Cryptography";
        
        // 备份当前值
        ProcessBuilder backupPb = new ProcessBuilder(
            "reg",
            "export",
            "HKLM\\" + registryPath,
            BACKUP_DIR + "\\MachineGuid_" + 
                LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")) + ".reg"
        );
        
        Process backupProcess = backupPb.start();
        if (backupProcess.waitFor() == 0) {
            System.out.println(GREEN + "[信息]" + RESET + " 已备份注册表值");
        }
        
        // 更新值
        ProcessBuilder pb = new ProcessBuilder(
            "reg",
            "add",
            "HKLM\\" + registryPath,
            "/v",
            "MachineGuid",
            "/t",
            "REG_SZ",
            "/d",
            guid,
            "/f"
        );
        
        Process p = pb.start();
        if (p.waitFor() != 0) {
            throw new Exception("更新注册表失败");
        }
        
        // 验证更新
        ProcessBuilder verifyPb = new ProcessBuilder(
            "reg",
            "query",
            "HKLM\\" + registryPath,
            "/v",
            "MachineGuid"
        );
        
        Process verifyProcess = verifyPb.start();
        if (verifyProcess.waitFor() == 0) {
            System.out.println(GREEN + "[信息]" + RESET + " 已成功更新注册表");
        }
    }

    /**
     * 显示操作结果
     */
    private static void displayResults(String machineId, String macMachineId,
                                     String devDeviceId, String sqmId) {
        System.out.println("\n" + GREEN + "[信息]" + RESET + " 已更新配置:");
        System.out.println(BLUE + "[调试]" + RESET + " machineId: " + machineId);
        System.out.println(BLUE + "[调试]" + RESET + " macMachineId: " + macMachineId);
        System.out.println(BLUE + "[调试]" + RESET + " devDeviceId: " + devDeviceId);
        System.out.println(BLUE + "[调试]" + RESET + " sqmId: " + sqmId);
    }

    /**
     * 处理自动更新设置
     */
    private static void handleAutoUpdate() throws Exception {
        System.out.println("\n" + YELLOW + "[询问]" + RESET + " 是否要禁用 Cursor 自动更新功能？");
        System.out.println("0) 否 - 保持默认设置");
        System.out.println("1) 是 - 禁用自动更新");

        Scanner scanner = new Scanner(System.in);
        String choice = scanner.nextLine();

        if ("1".equals(choice)) {
            String updaterPath = System.getenv("LOCALAPPDATA") + "\\cursor-updater";
            Files.write(Paths.get(updaterPath), new byte[0]);
            Files.setAttribute(Paths.get(updaterPath), "dos:readonly", true);
            System.out.println(GREEN + "[信息]" + RESET + " 已禁用自动更新");
        }
    }
}

File: src/com/moonciki/cursorsekiro/cursor/cursor_reset.py
import json
import os
import uuid
import shutil
import subprocess
from datetime import datetime
from pathlib import Path
import winreg
from typing import Dict, Optional, Tuple
from ..logger import Logger  # 添加Logger导入

class CursorReset:
    """用于重置Cursor编辑器标识符的工具类"""

    # 文件路径常量
    STORAGE_FILE = os.path.join(os.getenv('APPDATA'), 'Cursor', 'User', 'globalStorage', 'storage.json')
    BACKUP_DIR = os.path.join(os.getenv('APPDATA'), 'Cursor', 'User', 'globalStorage', 'backups')

    @staticmethod
    def create_backup_dir() -> None:
        """创建备份目录"""
        if not os.path.exists(CursorReset.BACKUP_DIR):
            os.makedirs(CursorReset.BACKUP_DIR)

    @staticmethod
    def backup_config() -> None:
        """备份现有配置文件"""
        if os.path.exists(CursorReset.STORAGE_FILE):
            Logger.info("正在备份配置文件...")
            backup_name = f"storage.json.backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            shutil.copy2(CursorReset.STORAGE_FILE, os.path.join(CursorReset.BACKUP_DIR, backup_name))

    @staticmethod
    def generate_machine_id() -> str:
        """生成新的机器ID"""
        prefix = "auth0|user_"
        prefix_hex = prefix.encode('utf-8').hex()
        random_hex = uuid.uuid4().hex
        return f"{prefix_hex}{random_hex[:64-len(prefix_hex)]}"

    @staticmethod
    def generate_ids() -> Dict[str, str]:
        """生成所有需要的ID"""
        ids_dict = {
            'machineId': CursorReset.generate_machine_id(),
            'macMachineId': str(uuid.uuid4()),
            'devDeviceId': str(uuid.uuid4()),
            'sqmId': f"{{{str(uuid.uuid4()).upper()}}}"
        }
        
        return ids_dict

    @staticmethod
    def update_machine_guid_reg() -> bool:
        """更新注册表中的MachineGuid"""
        try:
            registry_path = r"SOFTWARE\Microsoft\Cryptography"
            
            # 备份当前值
            try:
                with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, registry_path, 0, 
                                  winreg.KEY_READ) as key:
                    original_guid = winreg.QueryValueEx(key, "MachineGuid")[0]
                    Logger.info(f"当前MachineGuid: {original_guid}")
            except Exception:
                original_guid = None
                Logger.warn("无法读取当前MachineGuid")

            new_guid = str(uuid.uuid4())
            # 更新注册表
            with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, registry_path, 0, 
                              winreg.KEY_WRITE) as key:
                winreg.SetValueEx(key, "MachineGuid", 0, winreg.REG_SZ, new_guid)

            Logger.info(f"注册表更新成功: {new_guid}")
            return True

        except Exception as e:
            Logger.error(f"更新注册表失败: ", e)
            return False

    @staticmethod
    def update_config(new_ids: Dict[str, str]) -> bool:
        """更新配置文件"""
        try:
            if not os.path.exists(CursorReset.STORAGE_FILE):
                Logger.error("未找到配置文件")
                return False

            with open(CursorReset.STORAGE_FILE, 'r', encoding='utf-8') as f:
                config = json.load(f)

            # 更新配置
            config['telemetry.machineId'] = new_ids['machineId']
            config['telemetry.macMachineId'] = new_ids['macMachineId']
            config['telemetry.devDeviceId'] = new_ids['devDeviceId']
            config['telemetry.sqmId'] = new_ids['sqmId']

            # 写入新配置
            with open(CursorReset.STORAGE_FILE, 'w', encoding='utf-8') as f:
                json.dump(config, f, indent=2)

            Logger.info("配置文件更新成功")
            return True

        except Exception as e:
            Logger.error(f"更新配置失败: ", e)
            raise e

    @staticmethod
    def disable_auto_update() -> bool:
        """禁用Cursor的自动更新功能"""
        updater_path = os.path.join(os.getenv('LOCALAPPDATA'), 'cursor-updater')
        
        try:
            # 如果updater_path已经是文件且存在，说明已经禁用了自动更新
            if os.path.isfile(updater_path):
                Logger.info("已创建阻止更新文件，无需再次阻止")
                return True

            # 如果是目录，删除它
            if os.path.isdir(updater_path):
                shutil.rmtree(updater_path)
                Logger.info("成功删除 cursor-updater 目录")

            # 创建阻止文件
            with open(updater_path, 'w') as f:
                pass
            
            # 设置只读属性
            os.chmod(updater_path, 0o444)  # 设置为只读权限
            
            Logger.info("成功禁用自动更新")
            return True

        except Exception as e:
            Logger.error(f"禁用自动更新失败: ", e)
            Logger.warn("请手动操作：")
            Logger.info(f"1. 删除目录（如果存在）：{updater_path}")
            Logger.info(f"2. 创建空文件：{updater_path}")
            Logger.info("3. 将文件设置为只读")
            return False

    @staticmethod
    def reset_cursor(disable_update: bool = True) -> None:
        """
        执行Cursor重置流程
        
        Args:
            disable_update: 是否禁用自动更新，默认为False
        """
        try:
            # 创建备份目录
            CursorReset.create_backup_dir()
            
            # 备份当前配置
            CursorReset.backup_config()
            
            # 生成新ID
            new_ids = CursorReset.generate_ids()
            
            # 更新配置文件
            CursorReset.update_config(new_ids)
            # 更新注册表
            
            # 生成新GUID
            CursorReset.update_machine_guid_reg()
            Logger.info("Cursor重置完成")
            
            # 根据参数决定是否禁用自动更新
            if disable_update:
                Logger.info("正在禁用自动更新...")
                CursorReset.disable_auto_update()
            else:
                Logger.info("保持默认设置，不禁用自动更新")
            
                
        except Exception as e:
            Logger.error(f"发生未知错误: ", e)
            raise e

if __name__ == "__main__":
    
    CursorReset.reset_cursor()
